0016-consensus-parameters-change-proposal.md

```diff
@@ -0,0 +1,120 @@
+# ADR 0016: Consensus Parameters Change Proposal
+
+## Component
+
+Oasis Core
+
+## Changelog
+
+- 2022-09-15: Initial version
+
+## Status
+
+Proposed
+
+## Context
+
+Currently consensus parameters can only be changed with an upgrade governance
+proposal which is effective but not very efficient. Upgrades require downtime
+during which binaries need to be updated, nodes restarted and synced, consensus
+network version has to be increased etc. We would like to avoid this cumbersome
+procedure and change the parameters of a module as fast and as simple
+as possible without affecting the performance of the consensus layer.
+
+## Decision
+
+Implement governance proposal which changes consensus parameters only.
+
+## Implementation
+
+A new type of governance proposal named `ChangeParametersProposal` should be
+added to the consensus layer.
+
+```golang
+type ProposalContent struct {
+  ...
+  ChangeParameters *ChangeParametersProposal `json:"change_parameters,omitempty"`
+}
+```
+
+The proposal should contain:
+
+- the name of the consensus `Module` the changes should be applied to, and
+
+- a CBOR-encoded document `Changes` describing parameter changes.
+
+```golang
+// ChangeParametersProposal is a consensus parameters change proposal.
+type ChangeParametersProposal struct {
+  // Module identifies the consensus backend module to which changes should be
+  // applied.
+  Module string `json:"module"`
+  // Changes are consensus parameter changes that should be applied to
+  // the module.
+  Changes cbor.RawMessage `json:"changes"`
+}
+```
+
+If a `ChangeParametersProposal` closes as passed, the governance module will
+execute the proposal by broadcasting a message containing the proposal to other
+modules. Each module can then decide for itself whether to apply changes or not.
```

Comment:
Should there also be a message (dispatched to other modules) to validate the `Changes` when the proposal is submitted? As otherwise one could propose a bad value for the `Changes` field and it would only be discovered when the proposal is about to be executed (assuming the vote passed).

Also in case no module replies to the validation message, the proposal should be immediately rejected as not being supported.

Reply:
Not sure if we can nicely validate `Changes` at submission as parameters are not independent and can be changed by some other proposals that are in progress (e.g. `VotingPeriod` must be less than `UpgradeMinEpochDiff` so if one proposal increases the former and another decreases the latter, we can end up with invalid parameters even though validations passed). Maybe it is better to validate at execution (and maybe also at submission) and if validation fails, the module won't apply changes and the proposal will have to be re-submitted.

Will add immediate rejection if no module is listening to given changes, as early rejection has its benefits (e.g. funds are not deposited).

Reply:
Also at least things like whether a proposal is well-formed in some way should be checked immediately (e.g. it actually deserializes correctly to the expected format, parameters have the right ranges, etc.).

Agreed that execution could still fail later due to multiple proposals being executed and that validation should also be performed before execution. Make sure to describe how these things are handled in the ADR.